CRM/Civisocial/OAuthProvider/Facebook.php

```diff
@@ -0,0 +1,256 @@
+      //			email to continue) and then provide a link to re-authorize
+    }
+
+    // Authentication is successful. Fetch user profile
+    $userProfile = array();
+    if ($this->isAuthorized()) {
+      $userProfile = $this->getUserProfile();
+    }
+    else {
+      // Start over
+      CRM_Utils_System::redirect($this->getLoginUri());
+    }
+
+    $facebookUserId = CRM_Utils_Array::value("id", $userProfile);
+    $this->login($this->alias, $this->token, $facebookUserId);
+
+    if (!CRM_Civisocial_BAO_CivisocialUser::socialUserExists($facebookUserId, $this->alias)) {
+      $user = array(
+        'first_name' => CRM_Utils_Array::value("name", $userProfile),
+        'last_name' => '',
+        'display_name' => CRM_Utils_Array::value("name", $userProfile),
+        'preffered_language' => CRM_Utils_Array::value("lang", $userProfile),
+        'gender' => NULL,
+        'email' => CRM_Utils_Array::value("email", $userProfile),
+        'contact_type' => 'Individual',
+      );
+
+      // Create contact
+      $contactId = CRM_Civisocial_BAO_CivisocialUser::createContact($user);
+
+      // Create social user
+      $socialUser = array(
+        'contact_id' => $contactId,
+        'social_user_id' => $facebookUserId,
+        'access_token' => $this->token,
+      // @todo: Rename oauth_object in table to oauth_secret?
+        'backend' => $this->alias,
+        'created_date' => time(), // @todo: Created Date not being recorded
+      );
+
+      CRM_Civisocial_BAO_CivisocialUser::create($socialUser);
+    }
+
+    CRM_Core_Session::setStatus(ts('Login via Facebook successful.'), ts('Login Successful'), 'success');
+    // @todo: Is status shown on public pages?
+    CRM_Utils_System::redirect($requestOrigin);
+  }
+
+  /**
+   * Check if the user is connected to Facebook and authorized.
+   * It can also be used to validate access tokens after setting one.
+   *
+   * @return bool
+   */
+  public function isAuthorized() {
+    if ($this->token && isset($this->userProfile)) {
+      return TRUE;
+    }
+    $response = $this->get('me');
+    if (!$response) {
+      return FALSE;
+    }
+    $this->userProfile = $response;
+    return TRUE;
+  }
+
+  /**
+   * Check if all passed permissions have beeen granted
+   *
+   * @param array $permissions
+   *        Permissions to check if they have been granted
+   * @return array
+   *        An array of permissions that were denied
+   */
+  public function checkPermissions($permissions = array()) {
+    $grantedPermissions = $this->getGrantedPermissions();
+    if (count($permissions) > count($grantedPermissions)) {
+      return FALSE;
+    }
+    return array_diff($permissions, $grantedPermissions);
+  }
+
+  /**
+   * Get a list of granted permissions
+   *
+   * @return array | bool
+   *        FALSE if authorization fails
+   */
+  public function getGrantedPermissions() {
+    $response = $this->get('me/permissions');
```

Comment:
This node returns the list of permissions requested by the App and it's status i.e. whether it is granted.